tests/LexerTest.php

```diff
@@ -178,23 +178,97 @@ public function testBigNumbers()
         $this->assertEquals('922337203685477580700', $node->getValue());
     }
 
-    public function testStringWithEscapedDelimiter()
+    /**
+     * @dataProvider getStringWithEscapedDelimiter
+     */
+    public function testStringWithEscapedDelimiter(string $template, string $expected, array $expectedDeprecations = [])
     {
-        $tests = [
-            "{{ 'foo \' bar' }}" => 'foo \' bar',
-            '{{ "foo \" bar" }}' => 'foo " bar',
-        ];
-
         $lexer = new Lexer(new Environment(new ArrayLoader()));
-        foreach ($tests as $template => $expected) {
-            $stream = $lexer->tokenize(new Source($template, 'index'));
-            $stream->expect(Token::VAR_START_TYPE);
-            $stream->expect(Token::STRING_TYPE, $expected);
+        $stream = $lexer->tokenize(new Source($template, 'index'));
+        $stream->expect(Token::VAR_START_TYPE);
+        $stream->expect(Token::STRING_TYPE, $expected);
 
-            // add a dummy assertion here to satisfy PHPUnit, the only thing we want to test is that the code above
-            // can be executed without throwing any exceptions
-            $this->addToAssertionCount(1);
-        }
+        $this->assertSame($expectedDeprecations, $lexer->getDeprecations());
+    }
+
+    public function getStringWithEscapedDelimiter()
+    {
+        yield '{{ \'App\\\\Test\' }} => App\Test' => [
+            <<<'TWIG'
+            {{ 'App\\Test' }}
+            TWIG,
+            <<<'EOF'
+            App\Test
+            EOF,
+            [
+                'Character \ at position 5 does not require to be escaped.'
+            ]
+        ];
+        yield '{{ \'App\Test\' }} => App\Test' => [
+            <<<'TWIG'
+            {{ 'App\Test' }}
+            TWIG,
+            <<<'EOF'
+            App\Test
```

Comment:
This cannot be the behavior in a version *deprecating* the unnecessary escaping. The result of parsing must stay the same when introducing the deprecation (we should warn about a *future* behavior change, otherwise we are already breaking BC and the deprecation is useless).

Here is a good rule of thumb: except for the assertion on the deprecation message, there should be no difference in the tests between making them pass on the current stable 3.x release and the version containing the deprecation.

Reply:
Changed it back to the old behavior. The test now reflects what happens when you don't escape `\`.